cpp/src/arrow/array.cc

```diff
@@ -799,9 +807,8 @@ Status DictionaryArray::Transpose(MemoryPool* pool, const std::shared_ptr<DataTy
     default:
       return Status::NotImplemented("unexpected index type");
   }
-
-#undef TRANSPOSE_IN_OUT_CASE
 #undef TRANSPOSE_IN_CASE
+#undef TRANSPOSE_IN_OUT_CASE
 }
 
 // ----------------------------------------------------------------------
```

Comment:
I think `ValidateArray` should check `data()->dictionary` for dictionary arrays.